# Challenge 3

In this challenge we will work on the `Orders` data set. In your work you will apply the thinking process and workflow we showed you in Challenge 2.

You are serving as a Business Intelligence Analyst at the headquarter of an international fashion goods chain store. Your boss today asked you to do two things for her:

**First, identify two groups of customers from the data set.** The first group is **VIP Customers** whose **aggregated expenses** at your global chain stores are **above the 95th percentile** (aka. 0.95 quantile). The second group is **Preferred Customers** whose **aggregated expenses** are **between the 75th and 95th percentile**.

**Second, identify which country has the most of your VIP customers, and which country has the most of your VIP+Preferred Customers combined.**

## Q1: How to identify VIP & Preferred Customers?

We start by importing all the required libraries:

In [1]:
# import required libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore') 
pd.set_option('display.max_columns', None)

Next, import `Orders` from Ironhack's database into a dataframe variable called `orders`. Print the head of `orders` to overview the data:

In [2]:
# your code here
orders = pd.read_csv('Orders.csv')
orders.head(10)

Unnamed: 0  InvoiceNo StockCode  year  month  day  hour  \
0           0     536365    85123A  2010     12    3     8   
1           1     536365     71053  2010     12    3     8   
2           2     536365    84406B  2010     12    3     8   
3           3     536365    84029G  2010     12    3     8   
4           4     536365    84029E  2010     12    3     8   
5           5     536365     22752  2010     12    3     8   
6           6     536365     21730  2010     12    3     8   
7           7     536366     22633  2010     12    3     8   
8           8     536366     22632  2010     12    3     8   
9           9     536367     84879  2010     12    3     8   

                           Description  Quantity          InvoiceDate  \
0   white hanging heart t-light holder         6  2010-12-01 08:26:00   
1                  white metal lantern         6  2010-12-01 08:26:00   
2       cream cupid hearts coat hanger         8  2010-12-01 08:26:00   
3  knitted union flag hot water bottle         6  2010-12-01 08:26:00   
4       red woolly hottie white heart.         6  2010-12-01 08:26:00   
5         set 7 babushka nesting boxes         2  2010-12-01 08:26:00   
6    glass star frosted t-light holder         6  2010-12-01 08:26:00   
7               hand warmer union jack         6  2010-12-01 08:28:00   
8            hand warmer red polka dot         6  2010-12-01 08:28:00   
9        assorted colour bird ornament        32  2010-12-01 08:34:00   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34  
5       7.65       17850  United Kingdom         15.30  
6       4.25       17850  United Kingdom         25.50  
7       1.85       17850  United Kingdom         11.10  
8       1.85       17850  United Kingdom         11.10  
9       1.69       13047  United Kingdom         54.08

---

"Identify VIP and Preferred Customers" is the non-technical goal of your boss. You need to translate that goal into technical languages that data analysts use:

## How to label customers whose aggregated `amount_spent` is in a given quantile range?


We break down the main problem into several sub problems:

#### Sub Problem 1: How to aggregate the  `amount_spent` for unique customers?

#### Sub Problem 2: How to select customers whose aggregated `amount_spent` is in a given quantile range?

#### Sub Problem 3: How to label selected customers as "VIP" or "Preferred"?

*Note: If you want to break down the main problem in a different way, please feel free to revise the sub problems above.*

Now in the workspace below, tackle each of the sub problems using the iterative problem solving workflow. Insert cells as necessary to write your codes and explain your steps.

In [3]:
# your code here
# Primero algo de limpieza:
orders.drop(columns='Unnamed: 0',inplace = True)
orders.head()

InvoiceNo StockCode  year  month  day  hour  \
0     536365    85123A  2010     12    3     8   
1     536365     71053  2010     12    3     8   
2     536365    84406B  2010     12    3     8   
3     536365    84029G  2010     12    3     8   
4     536365    84029E  2010     12    3     8   

                           Description  Quantity          InvoiceDate  \
0   white hanging heart t-light holder         6  2010-12-01 08:26:00   
1                  white metal lantern         6  2010-12-01 08:26:00   
2       cream cupid hearts coat hanger         8  2010-12-01 08:26:00   
3  knitted union flag hot water bottle         6  2010-12-01 08:26:00   
4       red woolly hottie white heart.         6  2010-12-01 08:26:00   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34

In [4]:
orders.shape

(397924, 13)

In [5]:
# Vamos a ver quién gasta más

orders.groupby(by='CustomerID').sum().sort_values(by='amount_spent',ascending=False).head(10)

InvoiceNo      year  month    day   hour  Quantity  UnitPrice  \
CustomerID                                                                   
14646       1163267611   4182810  14191   6552  24488    197491    5176.09   
18102        243297801    866723   3746   1261   5587     64124    1940.92   
17450        188845149    677704   2292    842   4140     69993    1143.32   
16446          1688629      6033     22     11     27     80997       4.98   
14911       3196374868  11416155  46220  18930  68148     80515   26185.72   
12415        398543981   1439876   4254   2169   8061     77670    2097.08   
14156        780541003   2815397   8630   4284  17201     57885    5357.68   
17511        538173446   1936505   7215   2701  11320     64549    2221.28   
16029        134645942    486630   1777    748   2789     40208    8722.39   
12346           541431      2011      1      2     10     74215       1.04   

            amount_spent  
CustomerID                
14646          280206.02  
18102          259657.30  
17450          194550.79  
16446          168472.50  
14911          143825.06  
12415          124914.53  
14156          117379.63  
17511           91062.38  
16029           81024.84  
12346           77183.60

In [6]:
orders.year.max() - orders.year.min()   # Todas las compras han sido realizadas en un año, por lo que la fecha no es relevante
                                        # Si las compras se hubieran hecho en un periodo más largo, podría ser determinante
                                        # para clasificar a un VIP

1

In [7]:
orders.drop(columns=['year','month','day','hour','InvoiceDate'], inplace=True)
orders.head()

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   white hanging heart t-light holder         6   
1     536365     71053                  white metal lantern         6   
2     536365    84406B       cream cupid hearts coat hanger         8   
3     536365    84029G  knitted union flag hot water bottle         6   
4     536365    84029E       red woolly hottie white heart.         6   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34

In [8]:
orders_by_cust = orders.groupby(by='CustomerID').agg({'InvoiceNo':'count',
                                                      'Quantity':'sum',
                                                      'UnitPrice':'mean',
                                                      'amount_spent':'sum'}).sort_values(by='amount_spent',ascending=False)
orders_by_cust.shape

(4339, 4)

In [9]:
orders_by_cust.head(10)

InvoiceNo  Quantity  UnitPrice  amount_spent
CustomerID                                              
14646            2080    197491   2.488505     280206.02
18102             431     64124   4.503295     259657.30
17450             337     69993   3.392641     194550.79
16446               3     80997   1.660000     168472.50
14911            5677     80515   4.612598     143825.06
12415             716     77670   2.928883     124914.53
14156            1400     57885   3.826914     117379.63
17511             963     64549   2.306625      91062.38
16029             242     40208  36.042934      81024.84
12346               1     74215   1.040000      77183.60

In [10]:
orders_by_cust.describe().T

count         mean          std  min         25%         50%  \
InvoiceNo     4339.0    91.708689   228.792852  1.0   17.000000   41.000000   
Quantity      4339.0  1194.214335  5055.603049  1.0  160.000000  379.000000   
UnitPrice     4339.0     4.465257    34.207313  0.0    2.203233    2.917444   
amount_spent  4339.0  2053.793018  8988.248381  0.0  307.245000  674.450000   

                      75%        max  
InvoiceNo      100.000000    7847.00  
Quantity       993.000000  197491.00  
UnitPrice        3.829567    2033.10  
amount_spent  1661.640000  280206.02

Aquí hay iformación interesante:

- Quantity: Vemos que el 75% de los clientes ha comprado en 1 año 993 productos o menos
- UnitPrice: El 75% de los clientes compran productos con precio medio de 3.83 unidades monetarias. El max es 2033 u.m. Esta variable está muy desplazada a la derecha!!
- Amount_spent: el 75% de los clientes se gastan 1661,64 u.m o menos, el max está en 280206. Dos órdenes de magnitud más

Para seleccionar clientes VIP tenemos que ver perecentiles altos para estos tres campos.

In [11]:
# VAmos a empezar siendo estrictos: percentil 95. Cuantos clientes gastan mas que per95
amount_95 = np.percentile(orders_by_cust.amount_spent,95)
len(orders_by_cust[orders_by_cust.amount_spent > amount_95])

217

In [12]:
len(orders_by_cust[orders_by_cust.amount_spent > amount_95]) / len(orders_by_cust)*100

5.001152339248675

In [13]:
# Los clientes que gastan una cantidad superior al percentil 95, son solo 217 y representan el 5% de los clientes
# Vamos a compro
orders_by_cust[orders_by_cust.amount_spent > amount_95].describe().T

count          mean           std      min          25%  \
InvoiceNo     217.0    481.493088    844.426324     1.00   139.000000   
Quantity      217.0  11453.857143  19727.753122   121.00  3742.000000   
UnitPrice     217.0      5.856686     29.704734     0.40     2.403404   
amount_spent  217.0  20688.480230  35073.151824  5870.08  7044.680000   

                    50%           75%        max  
InvoiceNo      305.0000    460.000000    7847.00  
Quantity      5377.0000   8975.000000  197491.00  
UnitPrice        2.9675      3.750152     434.65  
amount_spent  9736.5200  16587.090000  280206.02

La conclusión que podemos obtener de la tabla describe anterior es, que los clientes que gastan más que el 95% compran mucha cantidad pero su precio unitario es bajo.
De hecho el percentil 75 de la tabla del 95 % es menor que la tabla con todos los clientes, pero la cantidad de percentil 75% es mayor en 1 orden de magnitud 1000 frente a 10000

Considero que se le puede asignar la categoría VIP a estos clientes por las razones previamente expuestas.

Voy a evaluar si los clientes que compran los productos unitarios más caros, también se les puede considerar clientes VIP

In [14]:
price_95 = np.percentile(orders_by_cust.UnitPrice,95)
len(orders_by_cust[orders_by_cust.UnitPrice > price_95])

217

Las casualidades no existen. Voy a comprobar que estos 217 clientes son los mismos que la tabla anterior

In [15]:
orders_by_cust[orders_by_cust.UnitPrice > price_95].index.sort_values() == orders_by_cust[orders_by_cust.amount_spent > amount_95].index.sort_values()
#Vale las caualidades existen 😂

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [16]:
orders_by_cust[(orders_by_cust.UnitPrice > price_95)&(orders_by_cust.amount_spent > amount_95)]  #Clientes VIP que han comprado productos con precio unitario > per95

InvoiceNo  Quantity   UnitPrice  amount_spent
CustomerID                                               
16029             242     40208   36.042934      81024.84
14096            5111     16352    6.521708      65164.79
15098               3       121  434.650000      39916.50
17857              54     19256   10.332407      26879.04
12744             222      5241   58.333288      21279.29
12678             165     10923    6.554667      17628.46
12536             261      2803   34.754521      12601.83
15502             213      4236   19.689953      10384.93
12590              68      4283   10.547353       9864.26
12757             371      2402   11.912642       7857.10
12980              20       886   15.904500       7374.90
15465              97      1257    6.698763       7354.62
12359             248      1622    8.620403       6372.58

In [17]:
orders_by_cust[(orders_by_cust.UnitPrice > price_95)].sort_values('UnitPrice', ascending=False).head(20)

InvoiceNo  Quantity    UnitPrice  amount_spent
CustomerID                                                
17846               1         1  2033.100000       2033.10
16742               1         1   464.900000        464.90
15802               1         1   451.420000        451.42
15098               3       121   434.650000      39916.50
18087               2      3907   426.810000       4055.72
13452               2         2   295.000000        590.00
15510               1         2   125.000000        250.00
15524               1         4   110.000000        440.00
18233               1         4   110.000000        440.00
17940              19      9755    98.528947       4106.44
15299               7      2948    88.515714       4507.01
15269               2        26    74.975000        408.80
14816               4         5    66.475000        271.85
14185               3         6    59.733333        197.65
12744             222      5241    58.333288      21279.29
16692               3       152    55.316667       1276.00
17500               5        17    54.300000       1138.00
13145               6       145    51.710000        618.52
16029             242     40208    36.042934      81024.84
12536             261      2803    34.754521      12601.83

Descartado incluir a los clientes que compran con precio unitario superior a per95, en la tabla anterior se puede observar que en su mayoría son compras únicas o poco frecuentes. Por lo tanto no son clientes asiduos

Vamos a proceder a establecer rango para clientes *Preferred*. Clientes cuyas compras estén entre el percentil 75-95

In [18]:
amount_75 = np.percentile(orders_by_cust.amount_spent,75)
len(orders_by_cust[(orders_by_cust.amount_spent > amount_75)&(orders_by_cust.amount_spent < amount_95)])

868

In [19]:
len(orders_by_cust[(orders_by_cust.amount_spent > amount_75)&(orders_by_cust.amount_spent < amount_95)]) / len(orders_by_cust)*100

20.0046093569947

Representa un 20% del total de clientes. Entre VIP y Preferred son el 25% de los clientes. Voy a ser más estricto, y voy a establecer que los clientes Preferred han de tener un acumulado de compras mayor que la media (La media es mayor que el percentil 75)

In [20]:
len(orders_by_cust[(orders_by_cust.amount_spent > orders_by_cust.amount_spent.mean()) & (orders_by_cust.amount_spent < amount_95)])

654

In [21]:
len(orders_by_cust[(orders_by_cust.amount_spent > orders_by_cust.amount_spent.mean()) & (orders_by_cust.amount_spent < amount_95)])/ len(orders_by_cust)*100

15.072597372666513

Guardamos en lista los clientes VIP y Preferred, y modificamos el dataset para añadir esta columna

In [22]:
vip = orders_by_cust[orders_by_cust.amount_spent > amount_95].index
preferred = orders_by_cust[(orders_by_cust.amount_spent > orders_by_cust.amount_spent.mean()) & (orders_by_cust.amount_spent < amount_95)].index
orders['Status'] = 'Regular'
orders.loc[orders.CustomerID.isin(vip),'Status'] = 'VIP'
orders.loc[orders.CustomerID.isin(preferred),'Status'] = 'Preferred'

In [23]:
orders.groupby(by='Status').agg({'InvoiceNo':'count',
                                'Quantity':'sum',
                                'UnitPrice':'mean',
                                'amount_spent':sum})

InvoiceNo  Quantity  UnitPrice  amount_spent
Status                                                 
Preferred     127073   1293388   2.980035   2166204.011
Regular       166367   1402821   2.941193   2255803.683
VIP           104484   2485487   3.560365   4489400.210

Now we'll leave it to you to solve Q2 & Q3, which you can leverage from your solution for Q1:

## Q2: How to identify which country has the most VIP Customers?

In [24]:
# your code here
vip_country = orders[orders.Status=='VIP'][['CustomerID','Country']].drop_duplicates() 
# Hay un customerID que tiene una entrada incorrecta. En esta consulta he obtenido 218 filas y deben ser 217
# Esto solo puede ser si un mismo CustomerID tiene dos entradas con paises distintos

In [25]:
vip_country

CustomerID         Country
26           12583          France
106          15311  United Kingdom
173          16029  United Kingdom
195          12431       Australia
209          17511  United Kingdom
...            ...             ...
277763       17509  United Kingdom
283681       12536          France
327149       12357     Switzerland
359337       18139  United Kingdom
393220       16000  United Kingdom

[218 rows x 2 columns]

In [26]:
ids = []
for fila in vip_country.itertuples():
    if fila[1] in ids:
        print(fila)
    else:
        ids.append(fila[1])

Pandas(Index=58027, CustomerID=12431, Country='Belgium')


In [27]:
vip_country[vip_country.CustomerID==12431]  #Este error no me afecta al resultado

CustomerID    Country
195         12431  Australia
58027       12431    Belgium

In [28]:
vip_country.Country.value_counts()

United Kingdom     177
Germany             10
France               9
Switzerland          3
Australia            2
EIRE                 2
Spain                2
Portugal             2
Japan                2
Singapore            1
Channel Islands      1
Finland              1
Belgium              1
Sweden               1
Cyprus               1
Netherlands          1
Norway               1
Denmark              1
Name: Country, dtype: int64

United Kingdom es el pais con más VIP

## Q3: How to identify which country has the most VIP+Preferred Customers combined?

In [29]:
# your code here
vip_prefeered_country = orders[(orders.Status=='VIP') | (orders.Status=='Preferred')][['CustomerID','Country']].drop_duplicates() 
vip_prefeered_country.Country.value_counts()


United Kingdom     739
Germany             34
France              27
Spain                9
Belgium              9
Switzerland          8
Portugal             6
Norway               6
Italy                4
Cyprus               4
Japan                4
Denmark              3
EIRE                 3
Austria              3
Australia            3
Finland              3
Sweden               2
Israel               2
Channel Islands      2
Iceland              1
Poland               1
Netherlands          1
Greece               1
Singapore            1
Canada               1
Name: Country, dtype: int64

United Kingdom es el pais con más VIP + Preferred